test/model/autocomplete_test.dart

```diff
@@ -971,6 +969,176 @@ void main() {
+        WildcardMentionOption() => query.testWildcardOption(candidate,
+          localizations: GlobalLocalizations.zulipLocalizations),
+        User() => query.testUser(candidate, (store ??= eg.store())),
+        UserGroup() => query.testUserGroup(candidate, (store ??= eg.store())),
+        _ => throw StateError('invalid candidate'),
+      };
+      return result?.rank;
+    }
+
+    void checkPrecedes(String query, Object a, Object b) {
+      check(rankOf(query, a)!).isLessThan(rankOf(query, b)!);
+    }
+
+    void checkSameRank(String query, Object a, Object b) {
+      check(rankOf(query, a)!).equals(rankOf(query, b)!);
+    }
+
+    test('wildcards, then users', () {
+      checkSameRank('', WildcardMentionOption.all, WildcardMentionOption.topic);
+      checkPrecedes('', WildcardMentionOption.topic, eg.user());
+      checkSameRank('', eg.user(), eg.user());
+    });
+
+    test('wildcard-vs-user more significant than match quality', () {
+      // Make the query an exact match for the user's name.
+      final user = eg.user(fullName: 'Ann');
+      checkPrecedes(user.fullName, WildcardMentionOption.channel, user);
+    });
+
+    test('user name matched case-insensitively', () {
+      final user1 = eg.user(fullName: 'Chris Bobbe');
+      final user2 = eg.user(fullName: 'chris bobbe');
+
+      checkSameRank('chris bobbe', user1, user2); // exact
+      checkSameRank('chris bo',    user1, user2); // total-prefix
+      checkSameRank('chr bo',      user1, user2); // word-prefixes
+    });
+
+    test('user name match: exact over total-prefix', () {
+      final user1 = eg.user(fullName: 'Chris');
+      final user2 = eg.user(fullName: 'Chris Bobbe');
+
+      checkPrecedes('chris', user1, user2);
+    });
+
+    test('user name match: total-prefix over word-prefixes', () {
+      final user1 = eg.user(fullName: 'So Many Ideas');
+      final user2 = eg.user(fullName: 'Some Merry User');
+
+      checkPrecedes('so m', user1, user2);
+    });
+
+    test('group name matched case-insensitively', () {
+      final userGroup1 = eg.userGroup(name: 'Mobile Team');
+      final userGroup2 = eg.userGroup(name: 'mobile team');
+
+      checkSameRank('mobile team', userGroup1, userGroup2); // exact
+      checkSameRank('mobile te',   userGroup1, userGroup2); // total-prefix
+      checkSameRank('mob te',      userGroup1, userGroup2); // word-prefixes
+    });
+
+    test('group name match: exact over total-prefix', () {
+      final userGroup1 = eg.userGroup(name: 'Mobile');
+      final userGroup2 = eg.userGroup(name: 'Mobile Team');
+
+      checkPrecedes('mobile', userGroup1, userGroup2);
+    });
+
+    test('group name match: total-prefix over word-prefixes', () {
+      final userGroup1 = eg.userGroup(name: 'So Many Ideas');
+      final userGroup2 = eg.userGroup(name: 'Some Merry Group');
+
+      checkPrecedes('so m', userGroup1, userGroup2);
+    });
+
+    test('email matched case-insensitively', () {
```

Comment:
The one other test that'd be good to have here is that emails are matched by prefix only: so e.g. check that "email@example.com" doesn't match a query "mail" or "example" or "example.com".